phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java

```diff
@@ -2213,56 +2213,229 @@ class PhoenixLevelRowCompactor {
+                    // There could be back to back delete family version markers and thus we need a loop to check it
+                    for (Cell deleteFamilyVersionCell : deleteFamilyVersionCellList) {
+                        if (cell.getTimestamp() > deleteFamilyVersionCell.getTimestamp()) {
+                            break;
+                        }
+                        if (cell.getTimestamp() == deleteFamilyVersionCell.getTimestamp()) {
+                            // It is masked
+                            if (cell.getType() != Cell.Type.Put) {
+                                if (cell.getTimestamp() == maxLookbackWindowStart) {
+                                    lastRowVersion.add(cell);
+                                } else if (!major) {
+                                    // Retain the delete markers if the compaction is not major
+                                    retainedCells.add(cell);
+                                }
+                            }
+                            cell = result.get(index + 1);
+                            if (!CellUtil.matchingColumn(cell, currentColumnCell)) {
+                                continue top;
+                            }
+                            index = index + 1;
+                        }
+                    }
+                }
+                if (deleteFamilyCell != null && deleteFamilyCell.getTimestamp() >= cell.getTimestamp()) {
+                    // This column is deleted by a delete family marker. Skip this column
+                    if (cell.getType() != Cell.Type.Put) {
+                        if (cell.getTimestamp() == maxLookbackWindowStart) {
+                            lastRowVersion.add(cell);
+                        } else if (!major) {
+                            // Retain the delete markers if the compaction is not major
+                            retainedCells.add(cell);
+                        }
+                    }
+                    index = skipColumn(result, currentColumnCell, retainedCells, index);
+                    continue top;
+                }
+                // Process back-to-back deleted cell versions. Phoenix currently does not use delete cell version
+                // markers. This processing should not happen and is added for completeness
+                while (cell.getType() == Cell.Type.Delete) {
+                    if (cell.getTimestamp() == maxLookbackWindowStart) {
                         lastRowVersion.add(cell);
+                    } else if (!major) {
+                        retainedCells.add(cell);
+                    }
+                    Cell nextCell = result.get(index + 1);
+                    if (!CellUtil.matchingColumn(currentColumnCell, nextCell)) {
+                        continue top;
+                    }
+                    if (nextCell.getType() == Cell.Type.Put && cell.getTimestamp() == nextCell.getTimestamp()) {
```

Comment:
We need to consume this delete cell so index should be incremented here (review by @tkhurana)